src/util/ConcurrentCache.h

```diff
@@ -0,0 +1,296 @@
+// Copyright 2021, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Johannes Kalmbach (kalmbacj@informatik.uni-freiburg.de)
+
+#ifndef QLEVER_CONCURRENTCACHE_H
+#define QLEVER_CONCURRENTCACHE_H
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <utility>
+
+#include "../util/HashMap.h"
+#include "./Synchronized.h"
+
+namespace ad_utility {
+
+using std::make_shared;
+using std::shared_ptr;
+
+/** This exception is thrown if we are waiting for a computation result,
+ * which is computed by a different thread and the computation in this
+ * other thread fails.
+ */
+class WaitedForResultWhichThenFailedException : public std::exception {
+  const char* what() const noexcept override {
+    return "Waited for a result from another thread which then failed";
+  }
+};
+
+// Implementation details, do not call them from outside this module.
+namespace ConcurrentCacheDetail {
+
+/**
+ * @brief A result of an expensive computation, that is only computed once
+ * @tparam Value - The result type of the computation.
+ *
+ * Usage: Multiple threads who wait for the same computational result of type
+ *        Value hold a pointer to this class. Exactly one of them actually
+ *        computes the result and supplies it via the finish() method, or calls
+ *        abort() to signal that the computation has failed. The other threads
+ *        may only call getResult(). This call blocks, until finish() or
+ *        abort() is called from the computing threads. If the result is
+ *        aborted, the call to getResult() will throw an
+ *        WaitedForResultWhichThenFailedException
+ *
+ *       This class is thread-safe
+ */
+template <typename Value>
+class ResultInProgress {
+ public:
+  ResultInProgress() = default;
+
+  // Distribute the computation results to all the threads that at some point
+  // have called or will call getResult(). Check that none of the other threads
+  // waiting for the result have already finished or were aborted.
+  void finish(shared_ptr<Value> result) {
+    std::lock_guard lockGuard(_mutex);
+    AD_CHECK(_status == Status::IN_PROGRESS);
+    _status = Status::FINISHED;
+    _result = std::move(result);
+    _cond_var.notify_all();
+  }
+
+  // Signal the failure of the computation to all the threads that at some point
+  // have called or will call getResult().
+  // If the total number of calls to finish() or abort() exceeds 1, the program
+  // will terminate.
+  void abort() {
+    AD_CHECK(_status == Status::IN_PROGRESS);
+    std::lock_guard lockGuard(_mutex);
+    _status = Status::ABORTED;
+    _cond_var.notify_all();
+  }
+
+  // Wait for another thread to finish the computation and obtain the result.
+  // If the computation is aborted, this function throws an
+  // AbortedInOtherThreadException
+  shared_ptr<const Value> getResult() {
+    std::unique_lock uniqueLock(_mutex);
+    _cond_var.wait(uniqueLock,
```

Comment:
I see no reason not to call it _condition_variable (it's used only in a few strategic places and in case of doubt always use the more verbose name, unless it makes the code unreadable because it appears so frequently or lines become super long)